### PR TITLE
Finished lazily initialized object refactor

### PR DESCRIPTION
**Pull request by bkonyi**

### CODE REVIEW

File: lib/src/base_impl.dart [hunk 1]
```diff
@@ -9,83 +9,41 @@ import 'dart:convert';
...
 }
 
-/// A base class for most DRAW objects which handles lazy-initialization of
-/// objects and Reddit API request state.
-abstract class RedditBase {
-  /// The current [Reddit] instance.
-  final Reddit reddit;
-
-  final RegExp _snakecaseRegExp = new RegExp("[A-Z]");
+abstract class RedditBaseInitializedMixin {
```
Comment (ckartik) [inline lib/src/base_impl.dart]: Would you mind confirming:
I looked into the Dart Docs (https://www.dartlang.org/guides/language/language-tour), there was no reference or use of hide for exporting classes. Is this a built-in language support for the PIMPL Idiom?
(Primary reason for asking is I noticed you removed public comments.)

File: lib/draw.dart [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
 library draw;
 
-export 'package:draw/src/reddit.dart';
+export 'package:draw/src/reddit.dart' hide RedditBaseInitializedMixin;
```
Comment (ckartik) [inline lib/draw.dart]: Would you mind confirming:
I looked into the Dart Docs (https://www.dartlang.org/guides/language/language-tour), there was no reference or use of `hide` for exporting classes. What does it do?
(Primary reason for asking is I noticed you removed a few public comments in "base_impl.dart")

File: lib/src/base_impl.dart [hunk 3]
```diff
@@ -9,83 +9,41 @@ import 'dart:convert';
...
-  /// The base request format for the current object.
-  String get infoPath => _infoPath;
-  String _infoPath;
+  Map _data = new Map();
 
   /// The fullname of a Reddit object.
   ///
   /// Reddit object fullnames take the form of 't3_15bfi0'.
-  Future<String> get fullname async => await property('name');
+  String get fullname => (data == null) ? null : data['name'];
```
Comment (ckartik) [inline lib/src/base_impl.dart]: Could we make a asynchronous property method in the base_impl for all of our getters? It would make the code more readable, since to re-write the check for (data == null), I've been doing this check for Multireddit.dart as well, and it seems all our models will likely need to do this.

File: lib/src/base_impl.dart [hunk 4]
```diff
@@ -9,83 +9,41 @@ import 'dart:convert';
...
-  final Reddit reddit;
-
-  final RegExp _snakecaseRegExp = new RegExp("[A-Z]");
+abstract class RedditBaseInitializedMixin {
+  Reddit get reddit;
+  String get infoPath;
 
   /// Returns the raw properties dictionary for this object.
   ///
   /// This getter returns null if the object is lazily initialized.
```
Comment (ckartik) [inline lib/src/base_impl.dart]: Isn't it returning a `new Map()` object?

File: lib/draw.dart (hunk 2 above)
Comment (bkonyi) [reply lib/draw.dart]: See docs for `hide` and `show` [here](https://www.dartlang.org/guides/language/language-tour#importing-only-part-of-a-library).

File: lib/src/base_impl.dart (hunk 4 above)
Comment (bkonyi) [reply lib/src/base_impl.dart]: Actually, I'm not sure why I'm allocating a map there since one should be set when we create an instance of `RedditBaseInitializedMixin`... I'll check this.

File: lib/src/models/comment_impl.dart [hunk 5]
```diff
@@ -60,7 +70,9 @@ class MoreComments extends RedditBase {
       : _children = data['children'],
         _count = data['count'],
         _parentId = data['parent_id'],
-        super.loadData(reddit, data);
+        super(reddit) {
+    setData(this, data);
```
Comment (ckartik) [inline lib/src/models/comment_impl.dart]: Since setData is a void function that is outside the scope of the super class (and abstract super class of that class), how are we able to call it here? Also what's with opening another scope after super, i.e the signature `SUPER LPAREN PARAM RPAREN LCURLY FUNC RCURLY` or in more concrete terms:
```
( 
super(reddit) {
     setData(this, data); 
}

File: lib/src/models/comment_impl.dart [hunk 6]
```diff
@@ -99,13 +111,17 @@ class MoreComments extends RedditBase {
   }
 
   Future<Comment> _loadComment(String commentId) async {
+    if (_submission is! Submission) {
+      _submission = await _submission.populate();
+    }
+    final submissionCast = _submission as Submission;
```
Comment (ckartik) [inline lib/src/models/comment_impl.dart]: What kind of voodoo is this?

Comment (ckartik) [inline lib/src/models/comment_impl.dart]: Style Comment: Are you sure we should keep Cast in the name of the variable, it makes it longer and hard to understand what's going on later on in the code when using that variable.

File: lib/src/models/comment_impl.dart [hunk 7]
```diff
@@ -139,25 +158,172 @@ class MoreComments extends RedditBase {
   }
 }
 
-class Comment extends CommentRef {
+/// A fully initialized class which represents a single Reddit comment.
+class Comment extends CommentRef
```
Comment (ckartik) [inline lib/src/models/comment_impl.dart]: It's probably just me, but I can't seem to find the constructor for this class.

File: lib/src/models/comment_impl.dart [hunk 8]
```diff
@@ -139,25 +158,172 @@ class MoreComments extends RedditBase {
...
+        VoteableMixin {
   // CommentModeration get mod; // TODO(bkonyi): implement
 
+  /// Has this [Comment] been approved.
+  bool get approved => data['approved'];
+
+  /// When this [Comment] was approved.
+  ///
+  /// Returns `null` if this [Comment] has not been approved.
+  DateTime get approvedAtUtc => (data['approved_at_utc'] == null)
```
Comment (ckartik) [inline lib/src/models/comment_impl.dart]: I may be sacrificing correctness for code readability, but would be better to have:
```
approved ? new DateTime.fromMillisecondsSinceEpoch(
           data['approved_at_utc'].round() * 1000,
           isUtc: true) : null;
```
Since we aren't checking to see if, data['approved_at_utc'] this might throw, so is it bad practice to rely on external logic (As we are here by relying on approved being true implying that 'approved_at_utc' will be a populated field) to not but the code execution into an internal exception.

File: lib/src/models/comment_impl.dart [hunk 9]
```diff
@@ -139,25 +158,172 @@ class MoreComments extends RedditBase {
...
+      ? null
+      : new DateTime.fromMillisecondsSinceEpoch(
+          data['approved_at_utc'].round() * 1000,
+          isUtc: true);
+
+  /// Which Redditor approved this [Comment].
+  ///
+  /// Returns `null` if this [Comment] has not been approved.
+  RedditorRef get approvedBy => (data['approved_by'] == null)
```
Comment (ckartik) [inline lib/src/models/comment_impl.dart]: Theoretical Question: (Disclaimer I don't really understand how in the world closures work, but I'm gonna ask a question that is probably going to sound stupid) If we use closures, does the new function created persist in the symbol table for the user session / or does it have to be generated again on every call to the getters?

In general though, I feel like we should have some sort of property extractor that is Synchronous in the base_impl.dart, because this null checking is getting out of hand (making getters which shouldn't have much logic, really hard to understand.

File: lib/src/models/comment_impl.dart [hunk 10]
```diff
@@ -139,25 +158,172 @@ class MoreComments extends RedditBase {
...
+
+  /// The author's flair text, if set.
+  ///
+  /// Returns `null` if the author does not have any flair text set.
+  String get authorFlairText => data['author_flair_text'];
+
+  /// When this [Comment] was removed.
+  ///
+  /// Returns `null` if the [Comment] has not been removed.
+  DateTime get bannedAtUtc => (data['banned_at_utc'] == null)
```
Comment (ckartik) [inline lib/src/models/comment_impl.dart]: Wanna have a Bool get banned/removed => (data['banned_at_utc'] != null).

Than replace the null check with a call to the other getter?

File: lib/src/models/comment_impl.dart [hunk 11]
```diff
@@ -139,25 +158,172 @@ class MoreComments extends RedditBase {
...
+      ? null
+      : new DateTime.fromMillisecondsSinceEpoch(
+          data['banned_at_utc'].round() * 1000,
+          isUtc: true);
+
+  /// Which Redditor removed this [Comment].
+  ///
+  /// Returns `null` if the [Comment] has not been removed.
+  RedditorRef get bannedBy =>
```
Comment (ckartik) [inline lib/src/models/comment_impl.dart]: This is again, hard to read. If only there was a better way! 😞

File: lib/src/models/comment_impl.dart [hunk 12]
```diff
@@ -139,25 +158,172 @@ class MoreComments extends RedditBase {
...
+      (data['banned_by'] == null) ? null : reddit.redditor(data['banned_by']);
```
Comment (ckartik) [inline lib/src/models/comment_impl.dart]: This is again, hard to read. We need to figure out a better way to do this! 😞

File: lib/src/models/comment_impl.dart (hunk 5 above)
Comment (bkonyi) [reply lib/src/models/comment_impl.dart]: `setData` is defined as a top-level method in `base_impl.dart`, so we can call that from here. Let me know if there's still something that isn't super clear (there's some interesting stuff going on here with libraries and scoping here that might not be obvious).

File: lib/src/models/comment_impl.dart (hunk 6 above)
Comment (bkonyi) [reply lib/src/models/comment_impl.dart]: Basically, `_submission` is a `SubmissionRef` so what we do here is ensure that `_submission` is populated. If it's not, `_submission` is an actual instance of `SubmissionRef` so we can populate it and get a `Submission` back. 

This is bad style on my part, I should have done this instead:

```dart
final Submission submissionCast = _submission;
```

Which accomplishes the same thing. If we were in C++, this is the same as me doing a cast to a `Submission` object.

Comment (bkonyi) [reply lib/src/models/comment_impl.dart]: Yeah, this is sloppy. I'll fix it. Thanks for calling me on it!

File: lib/src/models/comment_impl.dart (hunk 7 above)
Comment (bkonyi) [reply lib/src/models/comment_impl.dart]: It's wayyyyyyy below the properties :)

File: lib/src/models/comment_impl.dart (hunk 8 above)
Comment (bkonyi) [reply lib/src/models/comment_impl.dart]: That does look a bit cleaner, but the best thing to do here is probably just make a helper method (we do this type of operation a lot). Good catch.

File: lib/src/models/comment_impl.dart (hunk 9 above)
Comment (bkonyi) [reply lib/src/models/comment_impl.dart]: See above comment for making a helper for these sort of getters.

Not sure what the context is for the closure comment. Mind elaborating here or offline?

File: lib/src/models/comment_impl.dart (hunk 10 above)
Comment (bkonyi) [reply lib/src/models/comment_impl.dart]: See above comments. Probably should just use a helper function here.

File: lib/src/models/comment_impl.dart (hunk 12 above)
Comment (bkonyi) [reply lib/src/models/comment_impl.dart]: I'll add a helper to create these sort of things.

File: lib/src/models/comment_impl.dart (hunk 8 above)
Comment (bkonyi) [reply lib/src/models/comment_impl.dart]: Added helper.

File: lib/src/models/comment_impl.dart (hunk 9 above)
Comment (bkonyi) [reply lib/src/models/comment_impl.dart]: Added helper.

File: lib/src/models/comment_impl.dart (hunk 5 above)
Comment (ckartik) [reply lib/src/models/comment_impl.dart]: Do you have a name for the interesting behaviour, or at least some sort of string I could query google to learn more about it/docs reference?

Thanks!

File: lib/src/models/redditor.dart [hunk 13]
```diff
@@ -19,47 +19,93 @@ import 'package:draw/src/models/mixins/messageable.dart';
...
 
   /// The time the Redditor's account was created.
-  double get created => data['created'];
-
-  /// Returns the object's fullname.
-  ///
-  /// A fullname is an object's kind mapping (i.e., t3), followed by an
-  /// underscore and the object's ID (i.e., t1_c5s96e0).
-  Future<String> get fullname async => data['fullname'];
+  DateTime get createdUtc => (data['created_utc'] == null)
```
Comment (ckartik) [inline lib/src/models/redditor.dart]: Could GetterUtils be used here?

File: lib/src/models/redditor.dart [hunk 14]
```diff
@@ -19,47 +19,93 @@ import 'package:draw/src/models/mixins/messageable.dart';
...
+          isUtc: true);
 
   /// The amount of Reddit Gold a Redditor currently has.
   int get goldCreddits => data['gold_creddits'];
 
-  /// Whether the current Redditor is a Reddit employee.
+  /// The UTC date and time that the Redditor's Reddit Gold subscription ends.
+  ///
+  /// Returns `null` if the Redditor does not have Reddit Gold.
+  DateTime get goldExpiration => (data['gold_expiration'] == null)
```
Comment (ckartik) [inline lib/src/models/redditor.dart]: Could GetterUtils be used here?

File: lib/src/models/redditor.dart [hunk 15]
```diff
@@ -19,47 +19,93 @@ import 'package:draw/src/models/mixins/messageable.dart';
...
+
+  /// Redditor can see 18+ content.
+  bool get over18 => data['over_18'];
+
   /// Whether the Redditor has chosen to filter profanity.
   bool get preferNoProfanity => data['pref_no_profanity'];
 
-  Redditor.parse(Reddit reddit, Map data) : super.loadData(reddit, data) {
+  /// The date and time when the Redditor's suspension ends.
+  DateTime get suspensionExpirationUtc =>
```
Comment (ckartik) [inline lib/src/models/redditor.dart]: Could GetterUtils be used here?

File: lib/src/models/redditor.dart (hunk 13 above)
Comment (bkonyi) [reply lib/src/models/redditor.dart]: Yup, I totally forgot this file. Will do.

File: lib/src/models/redditor.dart (hunk 14 above)
Comment (bkonyi) [reply lib/src/models/redditor.dart]: Yup!

File: lib/src/models/redditor.dart (hunk 15 above)
Comment (bkonyi) [reply lib/src/models/redditor.dart]: Yup!

File: test/user/user_test.dart [hunk 16]
```diff
@@ -21,7 +21,7 @@ Future main() async {
     expect(me.commentKarma, equals(0));
     expect(me.linkKarma, equals(1));
     expect(me.goldCreddits, equals(0));
-    expect(me.created, equals(1501830779.0));
+    expect(me.createdUtc.millisecondsSinceEpoch / 1000, equals(1501830779));
```
Comment (ckartik) [inline test/user/user_test.dart]: Why not have:
`expect(me.createdUtc.millisecondsSinceEpoch, equals(1501830779000));`

File: test/user/user_test.dart [hunk 17]
```diff
@@ -76,10 +76,10 @@ Future main() async {
...
     final subreddit = subreddits[0];
     expect(subreddit.displayName, equals('drawapitesting'));
-    expect(await subreddit.isContributor, isTrue);
-    expect(await subreddit.isBanned, isFalse);
-    expect(await subreddit.title, equals('DRAW API Testing'));
-    expect(await subreddit.property('public_description'),
+    expect(subreddit.isContributor, isTrue);
+    expect(subreddit.isBanned, isFalse);
+    expect(subreddit.title, equals('DRAW API Testing'));
+    expect(subreddit.data['public_description'],
```
Comment (ckartik) [inline test/user/user_test.dart]: Does the use of `contains` here follow the specification or are we being lazy here?

File: test/user/user_test.dart [hunk 18]
```diff
@@ -76,10 +76,10 @@ Future main() async {
...
         contains('A subreddit used for testing'));
```
Comment (ckartik) [inline test/user/user_test.dart]: Does the use of `contains` here follow the specification or are we being lazy here?

File: lib/src/models/subreddit.dart [hunk 19]
```diff
@@ -276,18 +277,20 @@ class SubredditRef extends RedditBase
       }
       final currentIds = new Set();
       foundNewSubmission = false;
-      await for (final submission in search(query,
+      await for (var submission in search(query,
```
Comment (ckartik) [inline lib/src/models/subreddit.dart]: Why did you make `submission` mutable?

File: test/user/user_test.dart (hunk 16 above)
Comment (bkonyi) [reply test/user/user_test.dart]: The RHS is the value that's actually returned by the Reddit API, so this is just so that's clear.

File: test/user/user_test.dart (hunk 18 above)
Comment (bkonyi) [reply test/user/user_test.dart]: I felt that the actual description was a bit verbose and this still checks the same stuff. If you think I should put the full description, I can make that change.
